pkg/gadget-context/gadget-context.go

```diff
@@ -232,18 +236,116 @@ func (c *GadgetContext) SetParams(params []*api.Param) {
 	c.params = append(c.params, params...)
 }
 
-func (c *GadgetContext) SetMetadata(m []byte) {
+func (c *GadgetContext) SetMetadata(m []byte) error {
 	c.metadata = m
 
 	v := viper.New()
 	v.SetConfigType("yaml")
 	err := v.ReadConfig(bytes.NewReader(c.metadata))
 	if err != nil {
-		c.logger.Warn("unmarshalling metadata: %w", err)
-		return
+		return fmt.Errorf("unmarshalling metadata: %w", err)
 	}
 	c.logger.Debugf("loaded metadata as config")
+
+	// Extract custom params
+	params := make([]*api.Param, 0)
+
+	// Safeguard
+	if c.paramValues == nil {
+		c.paramValues = make(api.ParamValues)
+	}
+
+	paramValues := c.paramValues
+	customParams := v.Sub(customParamsKey)
+	if customParams != nil {
```

Comment:
I separated this logic and created some tests for it in https://github.com/inspektor-gadget/inspektor-gadget/commit/a5c0e22fc5353d2cc8426db74fd29b96f825bbee. There are a couple of tests that I think should fail.

Reply:
Thanks! We could add some plausibility checks at build time, but I'd like to postpone that a bit until we've seen how this feature will be adopted and extended.